feat(coderd/agentapi): implement sub agent api

DanielleMaywood opened the pull request:
Closes https://github.com/coder/internal/issues/619

Implement the `coderd` side of the AgentAPI for the upcoming dev-container agents work.

`agent/agenttest/client.go` is left unimplemented for a future PR working to implement the agent side of this feature.

DanielleMaywood committed:
chore: begin impl of dev container agent's `/coderd/agenetapi`

DanielleMaywood committed:
chore: add WorkspaceAgent rbac policy

DanielleMaywood committed:
chore: add Architecture and OperatingSystem to agent.proto

DanielleMaywood committed:
chore: remove todo

DanielleMaywood committed:
chore: appease linter

DanielleMaywood committed:
chore: add system context for test

DanielleMaywood committed:
chore: remove outdated code

DanielleMaywood committed:
chore: replace google.protobuf.Empty with DeleteDevContainerAgentResposne

DanielleMaywood committed:
chore: make test larger

This definitely isn't an ideal test. It feels quite large. Maybe I'll
try to extract the different things being tested and move them into
their own tests.

DanielleMaywood committed:
refactor: tests

Refactor the One Big Test into multiple separate unit tests

DanielleMaywood committed:
chore: appease linter

DanielleMaywood committed:
chore: begin validating agent name

Begin validation of the agent name. This uses the same logic (with the
exception of duplicate agent name detection) as the provisioner.

A future commit will aim to rectify the duplicate agent name issue.

DanielleMaywood committed:
chore: listen to feedback

mafredri commented on coderd/agentapi/devcontainer_agent.go:
```diff
@@ -0,0 +1,114 @@
+		ID:                       uuid.New(),
+		ParentID:                 uuid.NullUUID{Valid: true, UUID: parentAgent.ID},
+		CreatedAt:                a.Clock.Now(),
+		UpdatedAt:                a.Clock.Now(),
```
Suggestion: assigning to a variable creates an atomic entry where created/updated is the exact same. Sometimes useful to know if it's been changed or not.

mafredri commented on coderd/agentapi/devcontainer_agent.go:
```diff
@@ -0,0 +1,114 @@
+		AuthToken:                uuid.New(),
+		AuthInstanceID:           parentAgent.AuthInstanceID,
+		Architecture:             req.Architecture,
+		EnvironmentVariables:     pqtype.NullRawMessage{},
```
Note: I think this is correct, at least for now. If customers request inheritance we can look at it then, but ideally any envs needed are defined in `devcontainer.json`

DanielleMaywood replied:
Yeah we should definitely keep this in mind, or at least make sure to document that these variables aren't inherited.

mafredri commented on coderd/agentapi/devcontainer_agent.go:
```diff
@@ -0,0 +1,114 @@
+}
+
+func (a *DevContainerAgentAPI) CreateDevContainerAgent(ctx context.Context, req *agentproto.CreateDevContainerAgentRequest) (*agentproto.CreateDevContainerAgentResponse, error) {
+	//nolint:gocritic // We are the Dev Container Agent API, so this is safe.
```
This sounds a bit "trust me bruh", as a reader I'm still skeptical. 😂

mafredri commented on coderd/agentapi/devcontainer_agent_test.go:
```diff
@@ -0,0 +1,267 @@
+			require.Equal(t, agents[i].ID[:], agent.Id)
+			require.Equal(t, agents[i].Name, agent.Name)
+		}
+	})
```
I'd like to see some adversarial tests as well. I.e. what happens if I try to delete the parent agent, or someone else's subagent?

DanielleMaywood replied:
Will get to writing them!

mafredri commented on coderd/database/queries/workspaceagents.sql:
```diff
@@ -330,3 +330,9 @@ INNER JOIN workspace_resources ON workspace_resources.id = workspace_agents.reso
 WHERE workspace_builds.id = $1
 ORDER BY workspace_agent_script_timings.script_id, workspace_agent_script_timings.started_at;
+
+-- name: GetWorkspaceAgentsWithParentID :many
```
```suggestion
-- name: GetWorkspaceAgentsByParentID :many
```
I believe this is a bit more common nomenclature in coder/coder.

DanielleMaywood replied:
Can do!

mafredri reviewed (commented):
This largely looks good to me, nice work!

Just so I'm understanding this correctly. This permission model works because the `agentapi` is unique for each agent talking to it and you can't fake being another agent. Hence the simple queries which only filter on `id`/`parent_id`, etc. are sufficient. And on top of this, dbauthz verifies that agent is allowed to modify the workspace.

DanielleMaywood committed:
chore: rename GetWorkspaceAgentsWithParentID

DanielleMaywood committed:
chore: ensure createdAt and updatedAt match

DanielleMaywood commented on agent/proto/agent.proto:
```diff
@@ -391,4 +420,7 @@ service Agent {
 	rpc ReportConnection(ReportConnectionRequest) returns (google.protobuf.Empty);
+	rpc CreateDevContainerAgent(CreateDevContainerAgentRequest) returns (CreateDevContainerAgentResponse);
+	rpc DeleteDevContainerAgent(DeleteDevContainerAgentRequest) returns (DeleteDevContainerAgentResponse);
+	rpc ListDevContainerAgents(ListDevContainerAgentsRequest) returns (ListDevContainerAgentsResponse);
```
There was suggestion in the RFC to avoid `google.protobuf.Empty` in case we wanted to add fields in the future.

DanielleMaywood committed:
chore: rbac

DanielleMaywood committed:
chore: appease formatter

DanielleMaywood committed:
tests: refactor and add some adversarial tests

DanielleMaywood committed:
test: behavior is correct with no child agents

DanielleMaywood committed:
chore: appease linter

DanielleMaywood committed:
Merge branch 'main' into dm-devcontainer-agents-coderd

DanielleMaywood committed:
chore: update DRPCAgentClient25 interface

DanielleMaywood committed:
chore: rename sql parameter

DanielleMaywood committed:
Merge branch 'main' into dm-devcontainer-agents-coderd

DanielleMaywood committed:
Merge branch 'main' into dm-devcontainer-agents-coderd

DanielleMaywood committed:
chore: remove comment as recent PR resolves issue

mafredri reviewed (approved):
This looks largely good IMO. I suggest adding a bit more paranoia to the agent deletion query, but otherwise only minor suggestions and discussion points.

Pre-approving but deferring final approval to @spikecurtis and @johnstcn.

johnstcn commented on coderd/database/dbauthz/dbauthz.go:
```diff
@@ -319,6 +319,28 @@ var (
 		Scope: rbac.ScopeAll,
 	}.WithCachedASTValue()
 
+	subjectDevContainerAgentAPI = func(userID uuid.UUID, orgID uuid.UUID) rbac.Subject {
```
Out of scope of this PR, but there's an argument to be made for something like this:

```
type OrganizationID uuid.UUID
type UserID uuid.UUID
```

You could also have something like:
```
type UserOrgID struct {
  OrgID uuid.UUID
  UserID uuid.UUID
}
subjectDevContainerAgentAPI = func(UserOrgID) rbac.Subject { ... }
```

For the purposes of `AsDevcontainerAgentAPI`, you could maybe do something like the below, but it feels like overkill:

```
AsDevcontainerAgentAPI(ctx).InOrg(orgID).ForUser(userID)
```

We'll just need to be careful not to mix up OrgID and UserID.

johnstcn commented on coderd/agentapi/devcontainer_agent_test.go:
```diff
@@ -0,0 +1,408 @@
+					agent, err := api.Database.GetWorkspaceAgentByID(dbauthz.AsSystemRestricted(ctx), agentID) //nolint:gocritic // this is a test.
+					require.NoError(t, err)
+
+					require.Equal(t, tt.agentName, agent.Name)
+					require.Equal(t, tt.agentDir, agent.Directory)
+					require.Equal(t, tt.agentArch, agent.Architecture)
+					require.Equal(t, tt.agentOS, agent.OperatingSystem)
```
Suggest `assert.Equal` here so you get more information and not just the first failure.

johnstcn commented on coderd/agentapi/devcontainer_agent_test.go:
```diff
@@ -0,0 +1,408 @@
+			})
+
+			// Then: We expect it to fail and for the agent to still exist.
+			require.Error(t, err)
```
suggestion: check that we have the right error returned, I'd expect "not found"

johnstcn commented on coderd/database/queries/workspaceagents.sql:
```diff
@@ -330,3 +330,9 @@ INNER JOIN workspace_resources ON workspace_resources.id = workspace_agents.reso
 ORDER BY workspace_agent_script_timings.script_id, workspace_agent_script_timings.started_at;
+
+-- name: GetWorkspaceAgentsByParentID :many
+SELECT * FROM workspace_agents WHERE parent_id = @parent_id::uuid;
```
More paranoia: given that `parent_id` is nullable, what happens if someone calls this with `uuid.Nil`? Should we do a similar check for `AND parent_id IS NOT NULL`?

DanielleMaywood committed:
refactor: devcontainer agent -> subagent

Replace all naming of Dev Container Agent to Sub Agent.

DanielleMaywood committed:
refactor: deleteworkspaceagent to deleteworkspacesubagent

Refactor the DeleteWorkspaceAgentByID query to be
DeleteWorkspaceSubAgentByID and add a filter that only allows it to
delete an agent that has a non-NULL parent id.

DanielleMaywood committed:
fix: appease linter

DanielleMaywood committed:
test: paranoia

DanielleMaywood committed:
Merge branch 'main' into dm-devcontainer-agents-coderd

DanielleMaywood committed:
chore: apply feedback from spike

DanielleMaywood committed:
test: check err is a not authorized error

johnstcn commented on agent/proto/agent_drpc_old.go:
```diff
@@ -51,7 +51,17 @@ type DRPCAgentClient24 interface {
 
-// DRPCAgentClient25 is the Agent API at v2.5.
+// DRPCAgentClient25 is the Agent API at v2.5. It adds a ParentId field to the
+// agent manifest response. Compatible with Coder v2.23+
 type DRPCAgentClient25 interface {
 	DRPCAgentClient24
 }
+
```
It does seem odd to have an API version 2.5 that is equivalent to 2.4 with no changes. AFAICT to ship these changes with version 2.5 we would need to remove agent api version 2.5 from the current release branch. I'm not sure it's worth the hassle?

johnstcn replied:
Ah, then please disregard!

johnstcn commented on coderd/database/dbmem/dbmem.go:
```diff
@@ -1389,6 +1389,20 @@ func isDeprecated(template database.Template) bool {
 	return template.Deprecated != ""
 }
 
+func (q *FakeQuerier) DeleteWorkspaceAgentByID(ctx context.Context, id uuid.UUID) error {
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+
+	for i, agent := range q.workspaceAgents {
+		if agent.ID == id && agent.ParentID.Valid {
+			q.workspaceAgents = append(q.workspaceAgents[:i], q.workspaceAgents[i+1:]...)
+			return nil
+		}
+	}
+
+	return nil
+}
+
```
Is this needed?

DanielleMaywood replied:
Oops, have removed

DanielleMaywood committed:
fix: remove unused code

Accidentally left some unused code in dbmem.go
